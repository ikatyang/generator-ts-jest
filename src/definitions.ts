import dashify = require('lodash.kebabcase');
import Generator = require('yeoman-generator');

export interface Answers {
  project_name: string;
  project_description: string;
  project_keywords: any;
  github_username: string;
  user_name: string;
  user_email: string;
  tslint_config_preset: string;
  prettier_config_preset: string;
  node_version: string;
  source_directory: string;
  generated_directory: string;
  import_tslib: boolean;
  use_exact_version: boolean;
  enable_codecov: boolean;
  enable_greenkeeper: boolean;
}

export type Fields = Answers & {
  project_keywords: string[];
  github_profile: string;
  github_repository: string;
};

export const get_dependencies = (fields: Fields) =>
  fields.import_tslib ? ['tslib'] : [];

export const get_dev_dependencies = (fields: Fields) =>
  [
    '@types/jest',
    'jest',
    'standard-version',
    'ts-jest',
    'tslint',
    'typescript',
    'prettier',
    'tslint-plugin-prettier',
    'tslint-config-prettier',
  ]
    .concat(
      fields.tslint_config_preset.startsWith('tslint:')
        ? []
        : fields.tslint_config_preset,
    )
    .concat(
      fields.prettier_config_preset.length === 0
        ? []
        : fields.prettier_config_preset,
    );

export async function get_questions(generator: Generator) {
  const username = await get_username(generator);

  return {
    project_name: {
      type: 'input',
      message: 'Project Name',
      default: dashify(generator.appname),
    },
    project_description: {
      type: 'input',
      message: 'Project Description',
    },
    project_keywords: {
      type: 'input',
      message: 'Project Keywords',
    },
    user_name: {
      type: 'input',
      message: 'User Name',
      default: generator.user.git.name(),
    },
    user_email: {
      type: 'input',
      message: 'User Email',
      default: generator.user.git.email(),
    },
    github_username: {
      type: 'input',
      message: 'GitHub Username',
      default: username,
    },
    tslint_config_preset: {
      type: 'input',
      message: 'TSLint Config Preset',
      default: `tslint-config-${username}`,
    },
    prettier_config_preset: {
      type: 'input',
      message: 'Prettier Config Preset',
      default: `prettier-config-${username}`,
    },
    node_version: {
      type: 'list',
      message: 'Target Node Version',
      default: '4',
      choices: ['4', '6', '8'],
    },
    source_directory: {
      type: 'input',
      message: 'Source Directory',
      default: 'src',
    },
    generated_directory: {
      type: 'input',
      message: 'Generated Directory',
      default: 'lib',
    },
    import_tslib: {
      type: 'confirm',
      message: 'Import helpers from tslib',
      default: true,
    },
    use_exact_version: {
      type: 'confirm',
      message: 'Use Exact Version',
      default: true,
    },
    enable_codecov: {
      type: 'confirm',
      message: 'Enable Codecov',
      default: true,
    },
    enable_greenkeeper: {
      type: 'confirm',
      message: 'Enable Greenkeeper',
      default: true,
    },
  };
}

export const get_fields = (answers: Answers): Fields => {
  const keywords_text =
    typeof answers.project_keywords === 'string'
      ? answers.project_keywords.trim()
      : '';
  return {
    ...answers,
    project_keywords:
      keywords_text.length === 0
        ? []
        : keywords_text
            .split(/\s+/)
            .map(dashify)
            .sort(),
    github_profile: `https://github.com/${answers.github_username}`,
    github_repository: `https://github.com/${answers.github_username}/${
      answers.project_name
    }`,
  };
};

async function get_username(generator: Generator) {
  try {
    return generator.user.github.username();
  } catch {
    // istanbul ignore next
    return generator.user.git.email().split('@', 1)[0];
  }
}
