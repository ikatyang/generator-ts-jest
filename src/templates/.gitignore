coverage/
<%= generated_directory %>/
node_modules/
