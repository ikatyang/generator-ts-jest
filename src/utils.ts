import {execSync} from 'child_process';

export function get_tsconfig_target(node_version: string) {
  switch (node_version) {
    case '4':
      return 'es5';
    case '6':
    default:
      return 'es6';
  }
}

export function get_node_versions(node_version: string) {
  const targets: string[] = [];
  // tslint:disable:no-switch-case-fall-through
  switch (node_version) {
    case '4':
      targets.push('4');
    case '6':
    default:
      targets.push('6');
  }
  // tslint:enable:no-switch-case-fall-through
  return targets;
}

export function get_git_info(field: string): string {
  try {
    return execSync(`git config --get ${field}`, {encoding: 'utf8'}).trim();
  } catch (e) {
    return '';
  }
}
