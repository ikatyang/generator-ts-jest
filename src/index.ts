import * as fs from 'fs';
import * as Generator from 'yeoman-generator';
import {
  get_dependencies,
  get_questions,
  Answers,
  Fields,
} from './definitions';
import { get_node_versions, get_tsconfig_target } from './utils';

import dedent = require('dedent'); // tslint:disable-line:no-require-imports
import dashify = require('lodash.kebabcase'); // tslint:disable-line:no-require-imports

const helpers = {
  dedent,
  write_if_present: (write: boolean, content: string) => (write ? content : ''),
};

class TsJestGenerator extends Generator {
  public fields: Fields;

  // tslint:disable-next-line:promise-function-async
  public prompting() {
    return this.prompt(get_questions(this.appname)).then((answers: Answers) => {
      const keywords_text =
        typeof answers.project_keywords === 'string'
          ? answers.project_keywords.trim()
          : '';
      this.fields = {
        ...answers,
        project_keywords:
          keywords_text.length === 0
            ? []
            : keywords_text.split(/\s+/).map(dashify).sort(),
        tsconfig_target: get_tsconfig_target(answers.node_version),
        node_versions: get_node_versions(answers.node_version),
        github_profile: `https://github.com/${answers.github_username}`,
        github_repository: `https://github.com/${answers.github_username}/${answers.project_name}`,
      };
    });
  }

  public writing() {
    const template_dirname = `${__dirname}/templates`;
    const template_options = {
      ...helpers,
      ...this.fields,
    };
    fs
      .readdirSync(template_dirname)
      .filter(filename => filename !== 'src')
      .forEach(filename => {
        this.fs.copyTpl(
          this.templatePath(`${template_dirname}/${filename}`),
          this.destinationPath(filename.replace(/^-/, '.')),
          template_options,
        );
      });
    this.fs.copyTpl(
      this.templatePath(`${template_dirname}/src/index.ts`),
      this.destinationPath(`${this.fields.source_directory}/index.ts`),
      template_options,
    );
  }

  public install() {
    const dependencies = get_dependencies(this.fields);

    // istanbul ignore next
    this.yarnInstall(dependencies, {
      dev: true,
      ...this.fields.use_exact_version ? { exact: true } : {},
    });
  }
}

export = TsJestGenerator;
