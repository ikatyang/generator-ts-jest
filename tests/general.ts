import * as fs from 'fs';
import * as path from 'path';
import * as helpers from 'yeoman-test';
import {Answers} from '../src/definitions';
import MyGenerator = require('../src/index'); // tslint:disable-line:no-require-imports

let dirname: string;
let run_context: helpers.RunContext;

const original_cwd = process.cwd();
const default_answers: Answers = {
  project_name: 'my-awesome-project',
  project_description: 'my awesome description',
  project_keywords: 'my awesome keywords',
  github_username: 'ikatyang',
  user_name: 'Ika',
  user_email: 'ikatyang@gmail.com',
  tslint_config_preset: 'tslint-config-ikatyang',
  node_version: '6',
  source_directory: 'src',
  generated_directory: 'lib',
  use_exact_version: true,
  enable_codecov: true,
  enable_greenkeeper: true,
};

describe('default', () => {

  beforeAll(before_all(default_answers));
  afterAll(after_all());

  test('src/index.ts', () => {
    assert_file_content('src/index.ts');
  });
  test('.editorconfig', () => {
    assert_file_content('.editorconfig');
  });
  test('.gitattributes', () => {
    assert_file_content('.gitattributes');
  });
  test('.gitignore', () => {
    assert_file_content('.gitignore');
  });
  test('.npmignore', () => {
    assert_file_content('.npmignore');
  });
  test('.travis.yml', () => {
    assert_file_content('.travis.yml');
  });
  test('CHANGELOG.md', () => {
    assert_file_content('CHANGELOG.md');
  });
  test('jest.json', () => {
    assert_file_content('jest.json');
  });
  test('package.json', () => {
    assert_file_content('package.json');
  });
  test('README.md', () => {
    assert_file_content('README.md');
  });
  test('tsconfig.build.json', () => {
    assert_file_content('tsconfig.build.json');
  });
  test('tsconfig.json', () => {
    assert_file_content('tsconfig.json');
  });
  test('tslint.json', () => {
    assert_file_content('tslint.json');
  });
});

describe('project_keywords = ""', () => {
  beforeAll(before_all({...default_answers, project_keywords: ''}));
  afterAll(after_all());
  test('package.json', () => {
    assert_file_content('package.json');
  });
});
describe('node_version = 4', () => {
  beforeAll(before_all({...default_answers, node_version: '4'}));
  afterAll(after_all());
  test('.travis.yml', () => {
    assert_file_content('.travis.yml');
  });
  test('package.json', () => {
    assert_file_content('package.json');
  });
  test('tsconfig.json', () => {
    assert_file_content('tsconfig.json');
  });
});
describe('enable_codecov = false', () => {
  beforeAll(before_all({...default_answers, enable_codecov: false}));
  afterAll(after_all());
  test('.travis.yml', () => {
    assert_file_content('.travis.yml');
  });
  test('README.md', () => {
    assert_file_content('README.md');
  });
});
describe('enable_greenkeeper = false', () => {
  beforeAll(before_all({...default_answers, enable_greenkeeper: false}));
  afterAll(after_all());
  test('.travis.yml', () => {
    assert_file_content('.travis.yml');
  });
  test('README.md', () => {
    assert_file_content('README.md');
  });
});

function before_all(answers: Answers) {
  return () => {
    run_context = helpers
      .run(MyGenerator)
      .withPrompts(answers);
    return run_context.then(dir => { dirname = dir; });
  };
}

function after_all() {
  return () => {
    // run_context.cleanTestDirectory();
    process.chdir(original_cwd);
  };
}

function assert_file_content(filename: string, is_show = false) {
  const absolute_filename = path.resolve(dirname, filename);
  const file_content = fs.readFileSync(absolute_filename, 'utf8');
  if (is_show) {
    console.log(file_content); // tslint:disable-line:no-console
  } else {
    expect(file_content).toMatchSnapshot();
  }
}
