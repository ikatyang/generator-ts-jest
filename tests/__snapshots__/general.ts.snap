// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default .editorconfig 1`] = `
"root = true

[*]
indent_style = space
indent_size = 2
end_of_line = lf
insert_final_newline = true
trim_trailing_whitespace = true
"
`;

exports[`default .gitattributes 1`] = `
"* text eol=lf
*.png binary
"
`;

exports[`default .gitignore 1`] = `
"coverage/
lib/
node_modules/
"
`;

exports[`default .npmignore 1`] = `""`;

exports[`default .travis.yml 1`] = `
"language: node_js

node_js:
- 6
- 8

before_install:
- yarn global add codecov

env:
- JOB=test

script:
- if [ \\"$JOB\\" = \\"lint\\" ]; then yarn run lint; fi
- if [ \\"$JOB\\" = \\"test\\" ]; then yarn run test -- --verbose; fi
- if [ \\"$JOB\\" = \\"test-coverage\\" ]; then yarn run test -- --verbose --coverage; fi

after_script:
- if [ \\"$JOB\\" = \\"test-coverage\\" ]; then codecov; fi

cache:
  yarn: true
  directories:
  - node_modules

matrix:
  fast_finish: true
  include:
  - node_js: \\"stable\\"
    env: JOB=lint
  - node_js: \\"stable\\"
    env: JOB=test-coverage
"
`;

exports[`default CHANGELOG.md 1`] = `
"# Change Log

All changes to this project will be documented in this file.

> **Tags:**
> - 💥 [Breaking Change]
> - 🚀 [New Feature]
> - 🐛 [Bug Fix]
> - 📝 [Documentation]
> - 🏠 [Internal]
> - 💅 [Polish]

## Unreleased

#### 🚀 New Feature
- Release first version
"
`;

exports[`default README.md 1`] = `
"# my-awesome-project

[![npm](https://img.shields.io/npm/v/my-awesome-project.svg)](https://www.npmjs.com/package/my-awesome-project)
[![build](https://img.shields.io/travis/ikatyang/my-awesome-project/master.svg)](https://travis-ci.org/ikatyang/my-awesome-project/builds)
[![coverage](https://img.shields.io/codecov/c/github/ikatyang/my-awesome-project/master.svg)](https://codecov.io/gh/ikatyang/my-awesome-project)

my awesome description

[Changelog](https://github.com/ikatyang/my-awesome-project/blob/master/CHANGELOG.md)

## Install

\`\`\`sh
# using npm
npm install --save my-awesome-project

# using yarn
yarn add my-awesome-project
\`\`\`

## Usage

\`\`\`ts
// sample code
\`\`\`

## Development

\`\`\`sh
# lint
yarn run lint

# build
yarn run build

# test
yarn run test
\`\`\`

## License

MIT © [Ika](https://github.com/ikatyang)
"
`;

exports[`default jest.json 1`] = `
"{
  \\"testEnvironment\\": \\"node\\",
  \\"moduleFileExtensions\\": [\\"ts\\", \\"js\\", \\"json\\"],
  \\"testMatch\\": [\\"**/{tests,__tests__}/**/*.ts\\"],
  \\"transform\\": {\\"\\\\\\\\.ts$\\": \\"ts-jest/preprocessor\\"},
  \\"mapCoverage\\": true,
  \\"coverageReporters\\": [\\"lcov\\", \\"text-summary\\"],
  \\"collectCoverageFrom\\": [\\"src/**/*.ts\\"],
  \\"coverageThreshold\\": {
    \\"global\\": {
      \\"branches\\": 100,
      \\"functions\\": 100,
      \\"lines\\": 100,
      \\"statements\\": 100
    }
  }
}
"
`;

exports[`default package.json 1`] = `
"{
  \\"name\\": \\"my-awesome-project\\",
  \\"version\\": \\"0.0.0-dev\\",
  \\"description\\": \\"my awesome description\\",
  \\"keywords\\": [
    \\"awesome\\",
    \\"keywords\\",
    \\"my\\"
  ],
  \\"main\\": \\"lib/index.js\\",
  \\"types\\": \\"lib/index.d.ts\\",
  \\"repository\\": \\"https://github.com/ikatyang/my-awesome-project\\",
  \\"homepage\\": \\"https://github.com/ikatyang/my-awesome-project#readme\\",
  \\"author\\": {
    \\"name\\": \\"Ika\\",
    \\"email\\": \\"ikatyang@gmail.com\\",
    \\"url\\": \\"https://github.com/ikatyang\\"
  },
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"test\\": \\"jest --config ./jest.json\\",
    \\"lint\\": \\"tslint -p ./tsconfig.json --type-check\\",
    \\"prepublish\\": \\"yarn run build\\",
    \\"prebuild\\": \\"rm -rf ./lib\\",
    \\"build\\": \\"tsc -p tsconfig.build.json\\"
  },
  \\"devDependencies\\": {},
  \\"engines\\": {
    \\"node\\": \\">= 6\\"
  },
  \\"files\\": [
    \\"/lib/**/*\\"
  ]
}
"
`;

exports[`default src/index.ts 1`] = `""`;

exports[`default tsconfig.build.json 1`] = `
"{
  \\"extends\\": \\"./tsconfig.json\\",
  \\"include\\": [\\"src/index.ts\\"],
  \\"compilerOptions\\": {
    \\"rootDir\\": \\"src\\",
    \\"outDir\\": \\"lib\\"
  }
}
"
`;

exports[`default tsconfig.json 1`] = `
"{
  \\"compilerOptions\\": {
    \\"strict\\": true,
    \\"target\\": \\"es6\\",
    \\"module\\": \\"commonjs\\",
    \\"declaration\\": true
  },
  \\"include\\": [\\"src/**/*.ts\\"]
}
"
`;

exports[`default tslint.json 1`] = `
"{
  \\"extends\\": \\"tslint-config-ikatyang\\"
}
"
`;

exports[`node_version = 4 .travis.yml 1`] = `
"language: node_js

node_js:
- 4
- 6
- 8

before_install:
- yarn global add codecov

env:
- JOB=test

script:
- if [ \\"$JOB\\" = \\"lint\\" ]; then yarn run lint; fi
- if [ \\"$JOB\\" = \\"test\\" ]; then yarn run test -- --verbose; fi
- if [ \\"$JOB\\" = \\"test-coverage\\" ]; then yarn run test -- --verbose --coverage; fi

after_script:
- if [ \\"$JOB\\" = \\"test-coverage\\" ]; then codecov; fi

cache:
  yarn: true
  directories:
  - node_modules

matrix:
  fast_finish: true
  include:
  - node_js: \\"stable\\"
    env: JOB=lint
  - node_js: \\"stable\\"
    env: JOB=test-coverage
"
`;

exports[`node_version = 4 package.json 1`] = `
"{
  \\"name\\": \\"my-awesome-project\\",
  \\"version\\": \\"0.0.0-dev\\",
  \\"description\\": \\"my awesome description\\",
  \\"keywords\\": [
    \\"awesome\\",
    \\"keywords\\",
    \\"my\\"
  ],
  \\"main\\": \\"lib/index.js\\",
  \\"types\\": \\"lib/index.d.ts\\",
  \\"repository\\": \\"https://github.com/ikatyang/my-awesome-project\\",
  \\"homepage\\": \\"https://github.com/ikatyang/my-awesome-project#readme\\",
  \\"author\\": {
    \\"name\\": \\"Ika\\",
    \\"email\\": \\"ikatyang@gmail.com\\",
    \\"url\\": \\"https://github.com/ikatyang\\"
  },
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"test\\": \\"jest --config ./jest.json\\",
    \\"lint\\": \\"tslint -p ./tsconfig.json --type-check\\",
    \\"prepublish\\": \\"yarn run build\\",
    \\"prebuild\\": \\"rm -rf ./lib\\",
    \\"build\\": \\"tsc -p tsconfig.build.json\\"
  },
  \\"devDependencies\\": {},
  \\"engines\\": {
    \\"node\\": \\">= 4\\"
  },
  \\"files\\": [
    \\"/lib/**/*\\"
  ]
}
"
`;

exports[`node_version = 4 tsconfig.json 1`] = `
"{
  \\"compilerOptions\\": {
    \\"strict\\": true,
    \\"target\\": \\"es5\\",
    \\"module\\": \\"commonjs\\",
    \\"declaration\\": true
  },
  \\"include\\": [\\"src/**/*.ts\\"]
}
"
`;

exports[`project_keywords = "" package.json 1`] = `
"{
  \\"name\\": \\"my-awesome-project\\",
  \\"version\\": \\"0.0.0-dev\\",
  \\"description\\": \\"my awesome description\\",
  \\"keywords\\": [],
  \\"main\\": \\"lib/index.js\\",
  \\"types\\": \\"lib/index.d.ts\\",
  \\"repository\\": \\"https://github.com/ikatyang/my-awesome-project\\",
  \\"homepage\\": \\"https://github.com/ikatyang/my-awesome-project#readme\\",
  \\"author\\": {
    \\"name\\": \\"Ika\\",
    \\"email\\": \\"ikatyang@gmail.com\\",
    \\"url\\": \\"https://github.com/ikatyang\\"
  },
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"test\\": \\"jest --config ./jest.json\\",
    \\"lint\\": \\"tslint -p ./tsconfig.json --type-check\\",
    \\"prepublish\\": \\"yarn run build\\",
    \\"prebuild\\": \\"rm -rf ./lib\\",
    \\"build\\": \\"tsc -p tsconfig.build.json\\"
  },
  \\"devDependencies\\": {},
  \\"engines\\": {
    \\"node\\": \\">= 6\\"
  },
  \\"files\\": [
    \\"/lib/**/*\\"
  ]
}
"
`;
